tests/helpers.py

```diff
@@ -160,3 +166,174 @@ def __init__(self, **kwargs):
+            result += begin[:length] + "..." if len(begin) > (length + 3) else begin
+        result = result.replace("\n", "\\n").replace("\r", "\\r")
+        result += f" [{cls.__name__}"  # the other [ is added by pytest
+        return result
+
+    class TestClass:
+        """Test class for the generated tests."""
+
+        @pytest.mark.parametrize(
+            ("kwargs", "expected_bytes"),
+            parameters,
+            ids=tuple(generate_name(kwargs, i) for i, (kwargs, _) in enumerate(parameters)),
+        )
+        def test_serialization(self, kwargs: dict[str, Any], expected_bytes: bytes):
+            """Test serialization of the object."""
+            obj = cls(**kwargs)
+            serialized_bytes = obj.serialize()
+            assert serialized_bytes == expected_bytes, f"{serialized_bytes} != {expected_bytes}"
+
+        @pytest.mark.parametrize(
+            ("kwargs", "expected_bytes"),
+            parameters,
+            ids=tuple(generate_name(kwargs, i) for i, (kwargs, _) in enumerate(parameters)),
+        )
+        def test_deserialization(self, kwargs: dict[str, Any], expected_bytes: bytes):
+            """Test deserialization of the object."""
+            buf = Buffer(expected_bytes)
+            obj = cls.deserialize(buf)
+            assert cls(**kwargs) == obj, f"{cls.__name__}({kwargs}) != {obj}"
+            assert buf.remaining == 0, f"Buffer has {buf.remaining} bytes remaining"
+
+        @pytest.mark.parametrize(
+            ("kwargs", "exc"),
+            validation_fail,
+            ids=tuple(generate_name(kwargs, i) for i, (kwargs, _) in enumerate(validation_fail)),
+        )
+        def test_validation(self, kwargs: dict[str, Any], exc: type[Exception]):
+            """Test validation of the object."""
+            with pytest.raises(exc):
+                cls(**kwargs)
+
+        @pytest.mark.parametrize(
+            ("content", "exc"),
+            deserialization_fail,
+            ids=tuple(generate_name(content, i) for i, (content, _) in enumerate(deserialization_fail)),
+        )
+        def test_deserialization_error(self, content: bytes, exc: type[Exception]):
+            """Test deserialization error handling."""
+            buf = Buffer(content)
+            with pytest.raises(exc):
+                cls.deserialize(buf)
+
+    if len(parameters) == 0:
+        # If there are no serialization tests, remove them
+        del TestClass.test_serialization
+        del TestClass.test_deserialization
+
+    if len(validation_fail) == 0:
+        # If there are no validation tests, remove them
+        del TestClass.test_validation
+
+    if len(deserialization_fail) == 0:
+        # If there are no deserialization error tests, remove them
+        del TestClass.test_deserialization_error
+
+    # Set the names of the class
+    TestClass.__name__ = f"TestGen{cls.__name__}"
+
+    # Add the test functions to the global context
+    context[TestClass.__name__] = TestClass
```

Comment:
This feels so wrong, lol. I'm not sure how else to handle it, but I want to add a comment here just so I don't forget about it.

Reply:
Considering what I'm about to merge in #300 this has just gotten a lot more bearable lol. I guess we can keep it.